readme
(This is an old small project of my previous work.   
Today I'm upload it on Github for further maintainence (if I will).   
I'll add the content of description later... )

change_log.txt

ver		update
C01		initial
C01A	add "supportCharList"
		using require.js

ver		update
U00		initial
U00A	add "supportCharList"
		using require.js

c01a/le2ch.js
//var ga_ch_tab0=["啊",...
//var ga_ch_tab1=["０"...
/*
function le2chSwitchSpecialChar(iv){
	var rtv=iv;
	
	//switch ･(chinese ctr dot) and ｀,｀ is not used:
	if(iv==0xFF40) rtv=0xFF65;
	if(iv==0xFF65) rtv=0xFF40;
	
	return rtv;
}
*/
function le2chSolveQ2(ia_rangeLeft,ia_rangeRight,iv_char,iv_code,iv_c,iv_opt){
	var rtv;
	var lv=iv_code;
	
	if(iv_opt==0){
		var lv_zid=findInArray(ga_ch_tab0,iv_char);
		if(lv_zid==-1){return lv;}
		var lv_zidOut=solveQ2(0,ga_ch_tab0.length-1,lv_zid,iv_c);
		var lv_charOut=ga_ch_tab0[lv_zidOut];
		rtv=le2CharToCode(lv_charOut);
	}else if(iv_opt==1){
		var lv_zid=findInArray(ga_ch_tab01,iv_char);
		if(lv_zid==-1){return lv;}
		var lv_zidOut=solveQ2(0,ga_ch_tab01.length-1,lv_zid,iv_c);
		var lv_charOut=ga_ch_tab01[lv_zidOut];
		rtv=le2CharToCode(lv_charOut);
	}else{
		rtv=solveQ2(ia_rangeLeft,ia_rangeRight,lv,iv_c);
	}
	
	
	return rtv;
}

function le2chList(iv_opt){
	var rtv="";
	var i;
	var la_range;
	
	if(iv_opt==0){
		la_range=ga_ch_tab0
		for(i=0;i<ga_ch_tab0.length;i++){
			rtv=rtv+ga_ch_tab0[i];
		}
	}else if(iv_opt==1){
		la_range=ga_ch_tab01
		for(i=0;i<ga_ch_tab01.length;i++){
			rtv=rtv+ga_ch_tab01[i];
		}
	}else if(iv_opt==2){
		la_range=CT_RANGE[2];
		for(i=la_range[0];i<la_range[1];i++){
			rtv=rtv+le2CodeToChar(i);
		}
		la_range=CT_RANGE[3];
		for(i=la_range[0];i<la_range[1];i++){
			rtv=rtv+le2CodeToChar(i);
		}
		la_range=CT_RANGE[4];
		for(i=la_range[0];i<la_range[1];i++){
			rtv=rtv+le2CodeToChar(i);
		}
	}
	
	
	return rtv;
}



//----------------------------
//private:

c01a/le2ascii.js
var ga_ascii_tab0=["0","1","2","3","4","5","6","7","8","9","a","b","c","d","e","f","g","h","i","j","k","l","m","n","o","p","q","r","s","t","u","v","w","x","y","z","A","B","C","D","E","F","G","H","I","J","K","L","M","N","O","P","Q","R","S","T","U","V","W","X","Y","Z"];

function le2asciiSwitchSpecialChar(iv){
	var rtv=iv;
	
	//switch (space) and "`", "`" is not used:
	if(iv==0x20) rtv=0x60;
	if(iv==0x60) rtv=0x20;
	
	return rtv;
}

function le2asciiSolveQ2(ia_rangeLeft,ia_rangeRight,iv_char,iv_code,iv_c,iv_opt){
	var rtv;
	var lv=iv_code;
	
	if(iv_opt==0){
		var lv_zid=findInArray(ga_ascii_tab0,iv_char);
		if(lv_zid==-1){return lv;}
		var lv_zidOut=solveQ2(0,ga_ascii_tab0.length-1,lv_zid,iv_c);
		var lv_charOut=ga_ascii_tab0[lv_zidOut];
		rtv=le2CharToCode(lv_charOut);
		
	}else if(iv_opt==1){		
		rtv=solveQ2(ia_rangeLeft,ia_rangeRight,lv,iv_c);
	}
	
	
	return rtv;
}

function le2asciiList(iv_opt){
	var rtv="";
	var i;
	
	if(iv_opt==0){
		for(i=0;i<ga_ascii_tab0.length;i++){
			rtv=rtv+ga_ascii_tab0[i];
		}
	}else if(iv_opt==1){
		var la_range=CT_RANGE[1]
		for(i=la_range[0];i<la_range[1];i++){
			//skip not used char:
			if(i!=0x20 && i!=0x60);
				rtv=rtv+le2CodeToChar(i);
		}
	}
	
	
	return rtv;
}



c01a/le2devprint.js


//var C_COLOR_CAT_CJK_SYB="#00CCFF";

/*
var CT_RANGE = [
		new Array(-2,-1),			//(undefined)
		new Array(0x21,0x7E),		//ASCII
		new Array(0x4E00,0x9FCB),	//CJK
											//var C_RANGE_SUR=new Array(0xD800,0xDFFF);	//D800~DFFF
												//for High/Low Surrogates chars, should consider two bytes,
												//maybe enhance this later
		new Array(0xFF01,0xFF64),	//HFF
		new Array(0x3000,0x303F)	//CSP
A1E	];
*/

var CT_COLOR = [
		"gray",		//(undefined)
		"blue",	//ASCII
		"green",	//CJK
											//var C_RANGE_SUR=new Array(0xD800,0xDFFF);	//D800~DFFF
												//for High/Low Surrogates chars, should consider two bytes,
												//maybe enhance this later
		"red",		//HFF
		"orange"	//CSP
	];
//-----------------------------
//public
function le2devprint(iv_str){
	
	var lv_len;
	var lv_charCode;
	var i;
	
	lv_len=iv_str.length;
	fPrintDebugL(lv_len);
	
	for(i=0;i<lv_len;i++){
				
		lv_charCode=le2GetCharCodeAt(iv_str,i);
		
		le2devPrintCharCode(lv_charCode);
		le2devPrintChar(lv_charCode);
		fPrintDebug('\t');
	}
	fPrintDebugL('');
	
}

function le2devprintCJKfilter(iv_str){
	
	var la_range;
	var lv_len;
	var lv_charCode;
	var i;
	
	lv_len=iv_str.length;
	fPrintDebugL(lv_len);
	
	for(i=0;i<lv_len;i++){
				
		lv_charCode=le2GetCharCodeAt(iv_str,i);
		
		
		//locate range
		la_range=le2LocateRange(lv_charCode);
		if( la_range==CT_RANGE[2] ||
				  la_range==CT_RANGE[3] || 
				  la_range==CT_RANGE[4] 
			){					 
				le2devPrintCharCode(lv_charCode);
				le2devPrintChar(lv_charCode);
				
				fPrintDebug('\t');
				fPrintDebugL('');				
			}
			
	}
	
}
//-----------------------------
//private


function le2devPrintCharCode(iv_charCode){
	var lv;
	var lv_color;
	
	lv=iv_charCode;
	lv_color=CT_COLOR[le2devLocateRangeIndex(lv)];
	
	fPrintDebug('<font color="'+lv_color+'">'+utDecToHex(lv)+'</font> ');
}


function le2devLocateRangeIndex(iv_charCode){
	var rtv=0;
	var la;
	for(var i=0;i<CT_RANGE.length;i++){
		la=CT_RANGE[i];
		if( iv_charCode>=la[0] && iv_charCode<=la[1] ){
			rtv=i;
		}
	}
	
	return rtv;
}


function le2devPrintChar(iv_code){
	fPrintDebug(le2decodeCode(iv_code));	
}





c01a/le2.js
/*

ver		update
C01		initial
C01A	add "supportCharList"
		using require.js


*/
//-----------------------------
　require([
	'./c01a/ut', 
	'./c01a/math', 
	'./c01a/le2ascii', 
	'./c01a/le2ch_tab', 
	'./c01a/le2ch', 
	'./c01a/le2devprint'
]);
//const
var C_INIT_NOISE = "le2.js"
var CT_RANGE = [
		new Array(-2,-1),			// [0] (undefined)
		new Array(0x21,0x7E),		// [1] ASCII
		new Array(0x4E00,0x9FCB),	// [2] CJK
											//var C_RANGE_SUR=new Array(0xD800,0xDFFF);	//D800~DFFF
												//for High/Low Surrogates chars, should consider two bytes,
												//maybe enhance this later
		new Array(0xFF01,0xFF64),	// [3] HFF (2byte symbol, number, letter)
		new Array(0x3000,0x303F)	// [4] CSP (2byte chinese symbol)
	];

//-----------------------------
//public
function le2calStrC01(iv_key,iv_str,iv_optascii,iv_optch,iv_sp){
	var rtv="";
	var lv_p_key=0; //pointer of iv_key;
	var lv_char;
	var lv_code;
	var lv_codeOut;
	var lv_charOut;
	
	//gen key-to-use:
	var lv_noise=le2genNoise(iv_key);
	var lv_c;
	
	//get str length
	var lv_len=iv_str.length;
	
	//main loop:
	for(i=0;i<lv_len+iv_sp;i++){
		//pick and regenerate key
		lv_c=le2genKey(le2GetCodeAt(iv_key,lv_p_key),lv_noise,i);
		//adjust pointer
		lv_p_key++;
		if(lv_p_key>=iv_key.length) lv_p_key=0;
		
		
		if(i>=iv_sp){	//only when >= user inputted Start Pos
			//get char& char code
			lv_char=le2GetCharAt(iv_str,i-iv_sp);
			lv_code=le2GetCodeAt(iv_str,i-iv_sp);
			
			//calculate
			lv_codeOut=le2cal(lv_char,lv_code,lv_c,iv_optascii,iv_optch);
			
			//decode code to char
			lv_charOut=le2CodeToChar(lv_codeOut);
			
			rtv=rtv+lv_charOut;
			
		}
		
	}
		
	return rtv;
}

function le2suppCharList(iv_optascii,iv_optch){
	var rtv;
	
	lv_ascii=le2asciiList(iv_optascii);
	lv_ch=le2chList(iv_optch);
	
	rtv=lv_ascii+lv_ch;
	
	return rtv;
}
//-----------------------------
//private

function le2switchSpecialChar(iv){
	var rtv;
	
	//for ascii:
	rtv=le2asciiSwitchSpecialChar(iv);
	
	//for chinese:
	
	
	return rtv;
}

function le2locateRange(iv_code){
	var rtv=CT_RANGE[0];	//default value is set to undefined
	var la;
	for(var i=0;i<CT_RANGE.length;i++){
		la=CT_RANGE[i];
		if( iv_code>=la[0] && iv_code<=la[1] ){
			rtv=la;
		}
	}
	
	return rtv;
}

function le2cal(iv_char,iv_code,iv_c,iv_optascii,iv_optch){
	var rtv;
	var la_range;
	var lv_not_in_range=false;
	var lv=iv_code;

	//preprocess, switch special char if needed:
	lv=le2switchSpecialChar(lv);
	
	//locate range
	la_range=le2locateRange(lv);
	
	if(la_range==CT_RANGE[0]) lv_not_in_range=true;
	
	if(lv_not_in_range) rtv=lv;
	else {
		if(la_range==CT_RANGE[1]){ //ascii
			rtv=le2asciiSolveQ2(la_range[0],la_range[1],iv_char,lv,iv_c,iv_optascii);
		}else if( la_range==CT_RANGE[2] ||
				  la_range==CT_RANGE[3] || 
				  la_range==CT_RANGE[4] 
				 ){	//chinese
			rtv=le2chSolveQ2(la_range[0],la_range[1],iv_char,lv,iv_c,iv_optch);	
		}// code not in CT_RANGE is already marked "lv_not_in_range" 
		 // and don't need to be considered
		
	}
	
	//preprocess back, switch back special char:
	rtv=le2switchSpecialChar(rtv);
		
	return rtv;
}

function le2GetCharAt(iv_str,iv_p){
	return iv_str.charAt(iv_p);
}
function le2GetCodeAt(iv_str,iv_p){
	return le2CharToCode(iv_str.charAt(iv_p));
}

function le2CharToCode(iv_char){
	return iv_char.charCodeAt(0);	
}


function le2CodeToChar(iv){
	var u='%u'+utDecToHex(iv)
	return unescape(u);	
}

//-----------------------------
// func. that calculates the key:


function le2genNoise(iv){
	var lv;
	var rtv=0;
	
	/*
	//get the lower byte of the unicode value of 
	//each char in the string and xor them all with counter
	for(var i=0;i<iv.length;i++){
		lv=utLowByte(le2GetCodeAt(iv,i));
		fPrintDebugL("le2genNoise lv: "+lv);
		rtv=(rtv+i)^(lv*(i+1)); //xor
		fPrintDebugL("le2genNoise rtv: "+rtv);
	}*/
	
	rtv=BKDRHash(C_INIT_NOISE+iv);//https://segmentfault.com/a/1190000013132249
	
	fPrintDebugL("le2genNoise rtv: "+rtv);
	return rtv;
}

function le2genKey(iv_keyCode,iv_noise,iv_counter){
	var rtv=iv_keyCode;
	
	rtv=rtv^iv_noise;
	//fPrintDebugL("le2genKey rtv+iv_noise: "+rtv);
	
	rtv=rtv+iv_counter;
	//fPrintDebugL("le2genKey rtv+iv_counter: "+rtv);
	
	return rtv;
}
//-----------------------------
//ut
function utDecToHex(iv){
	var rtv=iv.toString(16).toUpperCase();
	if(rtv.length==1) rtv='0'+rtv;
	if(rtv.length==2) rtv='00'+rtv;
	return rtv;
}

/*function utLowByte(iv){
	var rtv;
	if(iv>256) rtv=iv%256;
	else rtv=iv;
	return rtv;
}*/

function BKDRHash(iv){
	/*pre definition, simulate long type*/
	//long = 4294967296
	var var_max = 32767; // = 0xFFFF / 2

	
	var seed = 131;
	var hash = 0;
	
	var hxs;
	
	for(var i=0;i<iv.length;i++){
		hxc=hash*seed;
		hxc=overflow(hxc,var_max);
		
		hash=hxc+iv.charCodeAt(i);
		hash=overflow(hash,var_max);
	}
	return hash;
}

function overflow(iv,max){
	var lv=iv;
	while(lv>max){
		lv=lv-max;
	}
	return lv;
}
//-----------------------------

//test



function decToHex(iv){return iv.toString(16);}




